File: backend/routers/published.py
from datetime import datetime, timezone

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy import select, desc
from sqlalchemy.ext.asyncio import AsyncSession

from database import get_db
from models import Publication, PublishAccount
from schemas import PublicationCreate, PublicationUpdate, PublicationOut

router = APIRouter(prefix="/published-articles", tags=["published"])

_STAT_FIELDS = {"read_count", "like_count", "look_count", "share_count"}


def _out(pub: Publication, account_name: str = "") -> PublicationOut:
    out = PublicationOut.model_validate(pub)
    out.account_name = account_name or ""
    return out


@router.get("", response_model=list[PublicationOut])
async def list_publications(
    status: str | None = None,
    account_id: str | None = None,
    db: AsyncSession = Depends(get_db),
):
    q = select(Publication, PublishAccount.name).join(
        PublishAccount, Publication.account_id == PublishAccount.id, isouter=True
    )
    if status:
        q = q.where(Publication.status == status)
    if account_id:
        q = q.where(Publication.account_id == account_id)
    q = q.order_by(desc(Publication.read_count), desc(Publication.created_at))
    rows = (await db.execute(q)).all()
    return [_out(pub, name) for pub, name in rows]


@router.post("", response_model=PublicationOut, status_code=201)
async def create_publication(body: PublicationCreate, db: AsyncSession = Depends(get_db)):
    pub = Publication(**body.model_dump())
    db.add(pub)
    await db.commit()
    await db.refresh(pub)
    acc = await db.get(PublishAccount, pub.account_id)
    return _out(pub, acc.name if acc else "")


@router.patch("/{pub_id}", response_model=PublicationOut)
async def update_publication(pub_id: int, body: PublicationUpdate, db: AsyncSession = Depends(get_db)):
    pub = await db.get(Publication, pub_id)
    if not pub:
        raise HTTPException(404, "发布记录不存在")
    data = body.model_dump(exclude_unset=True)
    for k in _STAT_FIELDS & set(data):
        if data[k] is not None and data[k] < 0:
            raise HTTPException(400, f"{k} 不能为负")
    if _STAT_FIELDS & {k for k, v in data.items() if v is not None}:
        pub.stats_as_of = datetime.now(timezone.utc)
    for k, v in data.items():
        setattr(pub, k, v)
    if data.get("status") == "published" and pub.published_at is None:
        pub.published_at = datetime.now(timezone.utc)
    await db.commit()
    await db.refresh(pub)
    acc = await db.get(PublishAccount, pub.account_id)
    return _out(pub, acc.name if acc else "")


@router.delete("/{pub_id}", status_code=204)
async def delete_publication(pub_id: int, db: AsyncSession = Depends(get_db)):
    pub = await db.get(Publication, pub_id)
    if not pub:
        raise HTTPException(404, "发布记录不存在")
    await db.delete(pub)
    await db.commit()

File: wemedia-studio/lib/api/published.ts
import { apiFetch } from './client'

export interface Publication {
  id: number
  draft_id: number
  account_id: string
  account_name: string
  platform: string
  title: string
  external_id: string
  url: string
  status: 'draft_box' | 'published'
  published_at: string | null
  read_count: number
  like_count: number
  look_count: number
  share_count: number
  stats_as_of: string | null
  created_at: string
  updated_at: string
}

export type PublicationPatch = Partial<Pick<Publication,
  'status' | 'url' | 'read_count' | 'like_count' | 'look_count' | 'share_count'>>

export function getPublications(params?: { status?: string; account_id?: string }): Promise<Publication[]> {
  const clean = Object.fromEntries(
    Object.entries(params ?? {}).filter(([, v]) => v !== undefined),
  ) as Record<string, string>
  const qs = new URLSearchParams(clean).toString()
  return apiFetch<Publication[]>(`/published-articles${qs ? `?${qs}` : ''}`)
}

export function updatePublication(id: number, patch: PublicationPatch): Promise<Publication> {
  return apiFetch<Publication>(`/published-articles/${id}`, {
    method: 'PATCH',
    body: JSON.stringify(patch),
  })
}

export function deletePublication(id: number): Promise<void> {
  return apiFetch<void>(`/published-articles/${id}`, { method: 'DELETE' })
}
